### PR TITLE
[tt-train] Add DP+TP support to train_nanogpt.py. Resolve problems with Python Llama.

### PR DESCRIPTION
**Pull request by imichalakTT**
### Summary

- Added new loudbox (dp=2; tp=4) llama 8b config.
- Made model_save_interval optional. Setting model_save_interval to 0 disables checkpointing.
- Added DP+TP support to `train_nanogpt.py`, matching behavior 1 to 1 with the C++ entry point -- nano_gpt and its DistributedLlama.
- Switched to the default LinearLayer random initalization in Llama, matching C++.
- Fixed sync_gradients memory leak.
- Disabled on-device init for `ttml.init.*` functions due to poor device RNG quality.

#### Python Llama (in TP mode) vs C++ DistributedLlama

Commands used:

```bash
python -u sources/examples/nano_gpt/train_nanogpt.py -c training_llama8b_dp2_tp4.yaml
```

```bash
../build_Debug/tt-train/sources/examples/nano_gpt/nano_gpt -c configs/training_configs/training_llama8b_dp2_tp4.yaml
```

> Early loss explosion on the Python side is caused by the token embeddings being randomly initialized with normal std=0.02 instead of C++'s std=1.0. 0.02 is recommended over 1.0 for Llama so I am leaving it. Can be easily solved by using a warmup LR scheduler.

losses

losses_diff

step_time

step_time_diff

#### Python Llama (in TP mode) vs C++ DistributedLlama (with tok_emb initialized with normally distributed random values with a standard deviation of 0.02 instead of 1.0)

losses

losses_diff

step_time

step_time_diff

##### Python

```bash
=== Memory Usage Summary ===

--- MODEL_CREATION ---
  DRAM: Segment Peak 3353.93 MB, Allocations 6675.62 MB, Deallocations 3329.44 MB, Segment Change +3346.18 MB
  DRAM: Cumulative Peak 3353.93 MB, Cumulative Current 3346.18 MB
  L1: Peak CB 1.50 MB, Peak Buffer 0.00 MB, Peak Total 1.50 MB
--- OPTIMIZER_CREATION ---
  DRAM: Segment Peak 6690.53 MB, Allocations 6690.53 MB, Deallocations 0.00 MB, Segment Change +6690.53 MB
  DRAM: Cumulative Peak 10036.71 MB, Cumulative Current 10036.71 MB
  L1: Peak CB 0.03 MB, Peak Buffer 0.00 MB, Peak Total 0.03 MB
--- FORWARD_PASS ---
  DRAM: Segment Peak 3435.62 MB, Allocations 7494.21 MB, Deallocations 4098.27 MB, Segment Change +3395.94 MB
  DRAM: Cumulative Peak 13472.33 MB, Cumulative Current 13432.65 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- BACKWARD_PASS ---
  DRAM: Segment Peak 72.12 MB, Allocations 21508.77 MB, Deallocations 24904.71 MB, Segment Change -3395.94 MB
  DRAM: Cumulative Peak 13504.78 MB, Cumulative Current 10036.71 MB
  L1: Peak CB 0.06 MB, Peak Buffer 0.00 MB, Peak Total 0.06 MB
--- FORWARD_PASS ---
  DRAM: Segment Peak 3435.62 MB, Allocations 7494.21 MB, Deallocations 4098.27 MB, Segment Change +3395.94 MB
  DRAM: Cumulative Peak 13504.78 MB, Cumulative Current 13432.65 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- BACKWARD_PASS ---
  DRAM: Segment Peak 72.12 MB, Allocations 21508.77 MB, Deallocations 24904.71 MB, Segment Change -3395.94 MB
  DRAM: Cumulative Peak 13504.78 MB, Cumulative Current 10036.71 MB
  L1: Peak CB 0.06 MB, Peak Buffer 0.00 MB, Peak Total 0.06 MB
--- FIRST_ITERATION_COMPLETE ---
  DRAM: Segment Peak 112.25 MB, Allocations 20072.02 MB, Deallocations 20071.88 MB, Segment Change +0.13 MB
  DRAM: Cumulative Peak 13504.78 MB, Cumulative Current 10036.85 MB
  L1: Peak CB 0.75 MB, Peak Buffer 0.00 MB, Peak Total 0.75 MB

=== Final Totals ===
Overall DRAM Peak: 13504.78 MB, Final DRAM Usage: 10036.85 MB
```

##### C++

```bash
=== Memory Usage Summary ===

--- MODEL_CREATION ---
  DRAM: Segment Peak 3353.93 MB, Allocations 6678.20 MB, Deallocations 3331.75 MB, Segment Change +3346.44 MB
  DRAM: Cumulative Peak 3353.93 MB, Cumulative Current 3346.44 MB
  L1: Peak CB 1.50 MB, Peak Buffer 0.00 MB, Peak Total 1.50 MB
--- OPTIMIZER_CREATION ---
  DRAM: Segment Peak 6691.03 MB, Allocations 6691.03 MB, Deallocations 0.00 MB, Segment Change +6691.03 MB
  DRAM: Cumulative Peak 10037.47 MB, Cumulative Current 10037.47 MB
  L1: Peak CB 0.03 MB, Peak Buffer 0.00 MB, Peak Total 0.03 MB
--- FORWARD_PASS ---
  DRAM: Segment Peak 3435.50 MB, Allocations 7493.77 MB, Deallocations 4097.89 MB, Segment Change +3395.88 MB
  DRAM: Cumulative Peak 13472.97 MB, Cumulative Current 13433.35 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- BACKWARD_PASS ---
  DRAM: Segment Peak 72.19 MB, Allocations 21508.13 MB, Deallocations 24904.00 MB, Segment Change -3395.88 MB
  DRAM: Cumulative Peak 13505.54 MB, Cumulative Current 10037.47 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- FORWARD_PASS ---
  DRAM: Segment Peak 3435.50 MB, Allocations 7493.77 MB, Deallocations 4097.89 MB, Segment Change +3395.88 MB
  DRAM: Cumulative Peak 13505.54 MB, Cumulative Current 13433.35 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- BACKWARD_PASS ---
  DRAM: Segment Peak 72.19 MB, Allocations 21508.13 MB, Deallocations 24904.00 MB, Segment Change -3395.88 MB
  DRAM: Cumulative Peak 13505.54 MB, Cumulative Current 10037.47 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- FIRST_ITERATION_COMPLETE ---
  DRAM: Segment Peak 112.25 MB, Allocations 20073.52 MB, Deallocations 20073.00 MB, Segment Change +0.51 MB
  DRAM: Cumulative Peak 13505.54 MB, Cumulative Current 10037.99 MB
  L1: Peak CB 0.75 MB, Peak Buffer 0.00 MB, Peak Total 0.75 MB

=== Final Totals ===
Overall DRAM Peak: 13505.54 MB, Final DRAM Usage: 10037.99 MB
```

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:imichalak/ttml-scripts)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:imichalak/ttml-scripts)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:imichalak/ttml-scripts)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:imichalak/ttml-scripts)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:imichalak/ttml-scripts)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:imichalak/ttml-scripts)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:imichalak/ttml-scripts)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:imichalak/ttml-scripts)

### CODE REVIEW

File: tt-train/sources/examples/nano_gpt/train_nanogpt.py [hunk 1]
```diff
@@ -248,6 +253,53 @@ def read_file_to_str(file_path: str) -> str:
...
+        if len(enabled_names) != 1:
+            raise ValueError(
+                f"Line mesh {shape} requires exactly one of enable_ddp / enable_tp; got enabled={enabled_names}"
+            )
+        active = nontrivial[0] if nontrivial else 0
+        axis_names[active] = enabled_names[0]
+    else:
+        if len(enabled_names) != n:
+            raise ValueError(f"2D mesh {shape} requires both axes assigned (DP and TP). Got enabled={enabled_names}")
+        # enabled_names is ordered ("dp", "tp") by construction above, matching
```
[the remark this answers is not in the document]
Comment (imichalakTT) [reply tt-train/sources/examples/nano_gpt/train_nanogpt.py:319]: I'm matching C++ behavior. Currently each dimension corresponds to one parallelism axis.

File: tt-train/configs/model_configs/llama8b_seq1024_tp.yaml
Comment (philei-tt) [inline tt-train/configs/model_configs/llama8b_seq1024_tp.yaml:1]: Why do we need this config? We usually target seq >= 2k, but it has wrong vocab size too

File: tt-train/configs/training_configs/training_llama8b_dp2_tp4.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  seed: 5489
+  # model_save_interval intentionally omitted: TP forbids --model_save_path / --resume,
+  # so checkpointing never fires for this config.
+  batch_size: 2  # = dp_size; 1 sample per DP shard per microbatch
+  gradient_accumulation_steps: 2  # effective global batch = batch_size * grad_acc = 4 (matches DP=4 numerics)
+  num_epochs: 50  # ~1075 samples/epoch at seq=1024 / batch_size=2 / grad_acc=2 -> ~268 optimizer steps/epoch; 5000 steps needs ~19 epochs
+  max_steps: 5000
+  use_clip_grad_norm: false  # TP-incompatible; both trainers reject use_clip_grad_norm=true under TP
+  clip_grad_norm_max_norm: 1.0
+  model_config: "${TT_METAL_RUNTIME_ROOT}/tt-train/configs/model_configs/llama8b_seq1024_tp.yaml"
```
Comment (philei-tt) [inline tt-train/configs/training_configs/training_llama8b_dp2_tp4.yaml]: Normal config with vocab 128k ans seq=2k should be fittable
In memory

File: tt-train/sources/examples/nano_gpt/train_nanogpt.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/examples/nano_gpt/train_nanogpt.py:319]: I think 3d mesh support isn't really working anyway 😢. There are a lot of asserts on 2d throughout metal infra and ttnn ops

File: tt-train/sources/examples/nano_gpt/train_nanogpt.py [hunk 3]
```diff
@@ -778,18 +855,27 @@ def sample_greedy(
...
 
     Args:
-        model_config: Universal model configuration
+        model_config: Universal model configuration.
+        use_tp: Whether the active mesh has a "tp" axis the model should
+            shard across. Forwarded to ``LlamaConfig.use_tp``.
 
     Returns:
-        A NanoGPT, Llama, or DeepSeek model instance
+        A NanoGPT, Llama, or DeepSeek model instance.
```
Comment (philei-tt) [inline tt-train/sources/examples/nano_gpt/train_nanogpt.py]: Qwen too btw

File: tt-train/sources/ttml/ttml/_mesh.py [hunk 4]
```diff
@@ -230,7 +230,8 @@ def sync_gradients(parameters, axis_names: tuple[str, ...] = ("dp",)):
     for _, param in parameters.items():
         if not param.is_grad_initialized():
             continue
-        grad_t = ttml.autograd.create_tensor(param.get_grad())
+        grad = param.get_grad()
         for axis in axes:
-            grad_t = ttml.ops.distributed.all_reduce(grad_t, True, axis)
-        param.set_grad(ttnn.multiply(grad_t.get_value(), inv_scaler))
+            grad = ttnn.all_reduce(grad, cluster_axis=axis)
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/_mesh.py:235]: Are you sure there's no performance difference with ttml's wrapper on all_reduce? I think it's using experimental.all_reduce_async + ttml manages it's own semaphores and other stuff.

File: tt-train/sources/ttml/ttml/init.py [hunk 5]
```diff
@@ -151,7 +151,8 @@ def uniform(a: float = 0.0, b: float = 1.0):
     """Uniform distribution over [a, b)."""
 
     def uniform_init(shape, mapper=None):
-        return ttml.ops.rand(shape, a, b, mesh_mapper=mapper)
+        data = np.random.uniform(low=a, high=b, size=tuple(shape)).astype(np.float32)
+        return ttml.autograd.Tensor.from_numpy(data, ttnn.Layout.TILE, ttnn.DataType.BFLOAT16, mapper)
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/init.py:155]: Is ttnn.rand causing accuracy issues here? If so, maybe we should open a ticket

File: tt-train/configs/model_configs/llama8b_seq1024_tp.yaml
Comment (imichalakTT) [reply tt-train/configs/model_configs/llama8b_seq1024_tp.yaml:1]: We don't. I'll remove this and swap model_config in training_llama8b_dp2_tp4.yaml from llama8b_seq1024_tp.yaml (vocab=30k, seq=1024) to llama8b.yaml (vocab=128k, seq=2048). New memory peak is
```
=== Final Totals ===
Overall DRAM Peak: 16973.59 MB, Final DRAM Usage: 10037.22 MB
```

File: tt-train/configs/training_configs/training_llama8b_dp2_tp4.yaml (hunk 2 above)
Comment (imichalakTT) [reply tt-train/configs/training_configs/training_llama8b_dp2_tp4.yaml]: Swapping

File: tt-train/sources/ttml/ttml/_mesh.py (hunk 4 above)
Comment (imichalakTT) [reply tt-train/sources/ttml/ttml/_mesh.py:235]: C++ uses ttnn::all_reduce for synchronizing gradients, but maybe it's worth switching to all_reduce_async.

```cpp
ttnn::Tensor synchronize_tensor(const ttnn::Tensor& tensor, const ttsl::SmallVector<uint32_t>& cluster_axes) {
    auto* device = &autograd::ctx().get_device();
    if (cluster_axes.size() == 0) {
        return tensor;
    }
    uint32_t scaler = 1U;
    for (const auto& cluster_axis : cluster_axes) {
        TT_FATAL(cluster_axis < device->shape().dims(), "Cluster axis must be within mesh shape");
        scaler *= device->shape()[cluster_axis];
    }
    if (scaler == 1U) {
        return tensor;
    }
    auto result = tensor;
    for (const auto& cluster_axis : cluster_axes) {
        result = ttnn::all_reduce(result, cluster_axis);
    }

    result = ttnn::multiply(result, 1.0F / static_cast<float>(scaler));
    return result;
}
```

File: tt-train/sources/ttml/ttml/init.py (hunk 5 above)
Comment (imichalakTT) [reply tt-train/sources/ttml/ttml/init.py:155]: Yes

Comment (imichalakTT) [reply tt-train/sources/ttml/ttml/init.py:155]: I tried looking for open issues related to rng, I couldn't find any related to the low entropy itself, but this issue talks about reproducibility and proposes some solutions:
https://github.com/tenstorrent/tt-metal/issues/35226

File: tt-train/sources/examples/nano_gpt/train_nanogpt.py (hunk 3 above)
Comment (imichalakTT) [reply tt-train/sources/examples/nano_gpt/train_nanogpt.py]: added mentions here: https://github.com/tenstorrent/tt-metal/pull/43254/changes/3ce23f128f651b1affe518f4dc0ada748f1fad62

File: tt-train/configs/training_configs/training_llama8b_dp2_tp4.yaml (hunk 2 above)
Comment (imichalakTT) [reply tt-train/configs/training_configs/training_llama8b_dp2_tp4.yaml]: fix here: https://github.com/tenstorrent/tt-metal/pull/43254/changes/71a52177de909f8e3e88306ec16bc0f7ad1d2444

File: tt-train/configs/model_configs/llama8b_seq1024_tp.yaml
Comment (imichalakTT) [reply tt-train/configs/model_configs/llama8b_seq1024_tp.yaml:1]: removed here: https://github.com/tenstorrent/tt-metal/pull/43254/changes/71a52177de909f8e3e88306ec16bc0f7ad1d2444

File: tt-train/sources/ttml/ttml/_mesh.py (hunk 4 above)
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/_mesh.py:235]: Oh I see. We use async version in TP communication. Not sure what's the difference. Thanks

**Review by philei-tt — APPROVED** [review]
